### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by ghislainbourgeois**
# Description

Initial implementation of the `httpreq` provider for ACME.

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      - name: Archive juju crashdump
+        if: failure()
+        uses: actions/upload-artifact@v3
+        with:
+          name: juju-crashdump
+          path: juju-crashdump-*.tar.xz
+
+  publish-charm:
+    name: Publish Charm
+    needs: integration-test
```
Comment (Gmerold) [inline .github/workflows/main.yaml:75]: Should we also require other test groups (ideally all of them) before we move on to the publishing? That's probably an edge case when something that was good in the PR fails after the merge, but @sanchezfdezjavier actually faced that problem recently.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+```bash
+juju relate httpreq-acme-operator:certificates <tls-certificates-requirer>
+````
+
+## Config
+
+### Required configuration properties
+
+- email: Let's Encrypt email address
```
Comment (Gmerold) [inline README.md:38]: And the `httpreq_endpoint`.

File: src/charm.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPReqAcmeOperatorCharm(AcmeClient):
+    """Main class that is instantiated every time an event occurs."""
+
+    REQUIRED_CONFIG = [
+        "HTTPREQ_ENDPOINT",
+    ]
```
Comment (Gmerold) [inline src/charm.py:23]: Do we need this? I can't find any usage for this list.

**Review by Gmerold — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yaml (hunk 1 above)
Comment (ghislainbourgeois) [reply .github/workflows/main.yaml]: Fixed, thanks!

File: README.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# httpreq-acme-operator
+
+## Description
+
+Let's Encrypt certificates in the Juju ecosystem for answering the DNS-01
+challenge through the HTTP Request plugin.
```
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply README.md]: Fixed, thanks!

File: README.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+## Config
+
+### Required configuration properties
+
+- email: Let's Encrypt email address
+
+### Optional configuration properties
+
+- server: Let's Encrypt server to use (default: `https://acme-v02.api.letsencrypt.org/directory`)
```
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply README.md:45]: Fixed, thanks!

File: README.md (hunk 2 above)
Comment (ghislainbourgeois) [reply README.md:40]: Fixed, thanks!
